clock_control: add vendor specific API to get HF clock startup time

bjarki-andreasen opened the pull request:
* Add vendor specific API to get HF clock startup time
* Clean up HF clock bindings (HFXO)
* Add HF clock bindings where missing

Note that for the nRF51, the clock was not called HFXO, but for simplicity its called HFXO in the devicetree

bjarki-andreasen-nordic committed:
drivers: clock_control: z_nrf_clock: add get_startup_time API

Implement vendor specific
z_nrf_clock_bt_ctlr_hf_get_startup_time_us() which gets the startup
time of the high frequency clock used for Bluetooth.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
dts: arm: nrf52: add hfxo binding and nodes

Add nrf52 series hfxo binding and nodes to socs.

Note that the values added to the soc .dtsi files are worst case
defaults, which will be replaced with optimal values at board level
in the future, as they depend on the specific crystal in use.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
dts: bindings: clock: nrf53-hfxo: add startup-time-us prop

Add the clock startup time property to the nrf53 series HFXO binding.

Note that the values added to the soc .dtsi files are worst case
defaults, which will be replaced with optimal values at board level
in the future, as they depend on the specific crystal in use.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
dts: bindings: clock: rename nordic,nrf-hfxo -> nordic,nrf54l-hfxo

Each nrf series has its own variant of the hfxo clock, specified
by the binding which includes the series in the name. The nrf54L
series is no different, hence the binding should be clearly
specified by having the series name within it.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
dts: bindings: clock: rename nordic,nrf-lfxo -> nordic,nrf54l-lfxo

Each nrf series has its own variant of the lfxo clock, specified
by the binding which includes the series in the name. The nrf54L
series is no different, hence the binding should be clearly
specified by having the series name within it.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
dts: bindings: clock: nrf54l-hfxo: add startup-time-us prop

Add the clock startup time property to the nrf54l series HFXO binding.

Note that the values added to the soc .dtsi files are worst case
defaults, which will be replaced with optimal values at board level
in the future, as they depend on the specific crystal in use.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
dts: arm: nrf51: add hfxo binding and nodes

Add nrf51 series hfxo binding and nodes to socs.

Note that the values added to the soc .dtsi files are worst case
defaults, which will be replaced with optimal values at board level
in the future, as they depend on the specific crystal in use.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

rugeGerritsen commented on dts/bindings/clock/nordic,nrf51-hfxo.yaml:
```diff
@@ -0,0 +1,17 @@
+  startup-time-us:
+    type: int
+    default: 1400
+    description: Startup time in microseconds.
```
We should document how to select this value. Similar for other added bindings

bjarki-andreasen replied:
Do you know how to select this value for the 51? I used https://github.com/nrfconnect/sdk-nrf/pull/22369/files which only mentions 52, 53 and 54

rugeGerritsen replied:
No, maybe @masz-nordic  knows?

rugeGerritsen commented on dts/bindings/clock/nordic,nrf54l-lfxo.yaml:
```diff
@@ -1,9 +1,9 @@
 # SPDX-License-Identifier: Apache-2.0
 
-description: Nordic nRF low-frequency crystal oscillator
+description: Nordic nRF low-frequency crystal oscillator  (nRF54L series)
```
```suggestion
description: Nordic nRF low-frequency crystal oscillator (nRF54L series)
```

rugeGerritsen commented on dts/bindings/clock/nordic,nrf51-hfxo.yaml:
```diff
@@ -0,0 +1,17 @@
+
+  startup-time-us:
+    type: int
+    default: 1400
```
Discussed offline. The bindings should not have default values as these are board specific. That will require a board designer to specify what is being used.

Decided to leave that for a followup

rugeGerritsen reviewed (commented):
@cvinayak , after this change EVENT_OVERHEAD_XTAL_US could start using the device-tree defined startup latency

cvinayak commented on dts/bindings/clock/nordic,nrf54l-hfxo.yaml:
```diff
@@ -1,16 +1,21 @@
 
 properties:
   clock-frequency:
     const: 32000000
```
128M output HCLK?

bjarki-andreasen commented on dts/bindings/clock/nordic,nrf51-hfxo.yaml:
```diff
@@ -0,0 +1,17 @@
+
+properties:
+  clock-frequency:
+    const: 32000000
```
done

bjarki-andreasen commented on dts/bindings/clock/nordic,nrf53-hfxo.yaml:
```diff
@@ -11,6 +11,14 @@ properties:
   clock-frequency:
     const: 32000000
```
outside scope of this PR

bjarki-andreasen-nordic committed:
drivers: clock_control: nrf: adapt NRF_HFINT_CALIBRATION option

CLOCK_CONTROL_NRF_HFINT_CALIBRATION depended on the renamed
nordic,nrf-hfxo -> nordic,nrf54l-hfxo. Update config to depend on
DT_HAS_NORDIC_NRF54L_HFXO_ENABLED to match new compat name.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

rugeGerritsen commented on dts/bindings/clock/nordic,nrf52-hfxo.yaml:
```diff
@@ -0,0 +1,30 @@
+  startup-time-us:
+    type: int
+    description: |
+      Startup time in microseconds. It corresponds to the twice the time
```
The startup time does not correspond to twice the time between TASKS_HFCLKSTART and EVENTS_HFCLKSTARTED, but this is a way of obtaining a safe approximation of the value.

I would suggest phrasing it like:
```suggestion
      Startup time in microseconds
      The value can be obtained by first measuring the time between
      TASKS_HFCLKSTART and EVENTS_HFCLKSTARTED. Then multiply
      the value by 2 to account for temperature and supply variations.
      
```

cvinayak commented:
> @cvinayak , after this change EVENT_OVERHEAD_XTAL_US could start using the device-tree defined startup latency

@rugeGerritsen PR to cover the use of devicetree in bsim CI(not testing on target today, will test in the future at the next opportunity).
https://github.com/zephyrproject-rtos/zephyr/pull/90665

cvinayak commented:
@aescolar do we need any DTS updates for bsim boards to pick the node value?

cvinayak commented:
> @aescolar do we need any DTS updates for bsim boards to pick the node value?

Looks like bsim CI already picks it up :-) .. there is mesh test failing.

bjarki-andreasen commented on soc/nordic/nrf52/soc.c:
```diff
@@ -41,6 +46,7 @@ static int nordicsemi_nrf52_init(void)
 	nrf_power_dcdcen_vddh_set(NRF_POWER, true);
 #endif
 
+	NRF_CLOCK->HFXODEBOUNCE = DT_PROP(DT_NODELABEL(hfxo), debounce_us) / 16;
```
Looked into it, the default debounce time is accounted for as it is part of the time between TASKS_HFCLKSTART and EVENTS_HFCLKSTARTED so don't even have to mention it. Though it is still a feature we need to add IMO.

bjarki-andreasen replied:
done
